calico_kubernetes/policy.py

```diff
@@ -0,0 +1,224 @@
+"""
+Usage: parse (allow|deny) [(
+      (tcp|udp) [(from [(ports <SRCPORTS>)] [(label <SRCLABEL>)] [(cidr <SRCCIDR>)])]
+                [(to   [(ports <DSTPORTS>)] [(label <DSTLABEL>)] [(cidr <DSTCIDR>)])] |
+      icmp [(type <ICMPTYPE> [(code <ICMPCODE>)])]
+           [(from [(label <SRCLABEL>)] [(cidr <SRCCIDR>)])]
+           [(to   [(label <DSTLABEL>)] [(cidr <DSTCIDR>)])] |
+      [(from [(label <SRCLABEL>)] [(cidr <SRCCIDR>)])]
+      [(to   [(label <DSTLABEL>)] [(cidr <DSTCIDR>)])]
+    )]
+"""
+import re
+import logging
+from netaddr import IPNetwork, AddrFormatError
+from docopt import docopt
+from pycalico.datastore_datatypes import Rule
+
+_log = logging.getLogger(__name__)
+
+
+class PolicyParser(object):
+    """
+    Class which parses Kubernetes annotation-based policy strings and
+    converts them to libcalico Rules objects.
+
+    Usage:
+      parser = PolicyParser(namespace)
+      rule = parser.parse_line("allow tcp from label stage=prod")
+    """
+    def __init__(self, namespace):
+        self.namespace = namespace
+
+    def parse_line(self, policy):
+        """
+        Takes a single line of policy as defined in the annotations of a
+        pod and returns the equivalent libcalico Rule object.
+        :param policy: Policy string from annotations.
+        :return: A Rule object which represent the given policy.
+        """
+        _log.info("Parsing policy line: '%s'", policy)
+        splits = policy.split()
+
+        try:
+            args = docopt(__doc__, argv=splits)
+        except BaseException:
+            raise PolicyParseError("Failed to parse policy: %s", policy)
+
+        # Generate a rule object from the arguments.
+        rule = self._generate_rule(args)
+
+        return rule
+
+    def _generate_rule(self, arguments):
+        """
+        Generates a libcalico
+        :param arguments: A dictionary of arguments with the following format.
+
+        {
+            "<DSTCIDR>": null,
+            "<DSTPORTS>": null,
+            "<DSTTAG>": null,
+            "<ICMPCODE>": null,
+            "<ICMPTYPE>": null,
+            "<SRCCIDR>": null,
+            "<SRCPORTS>": "80",
+            "<SRCTAG>": null,
+            "allow": true,
+            "cidr": 0,
+            "code": false,
+            "deny": false,
+            "from": true,
+            "icmp": false,
+            "ports": 1,
+            "tag": 0,
+            "tcp": true,
+            "to": false,
+            "type": false,
+            "udp": false
+        }
+
+        :return: A libcalico Rule object
+        """
+        # We only support whitelist rules.
+        rule_args = {"action": "allow"}
+
+        # Get arguments.
+        if arguments.get("tcp"):
+            protocol = "tcp"
+        elif arguments.get("udp"):
+            protocol = "udp"
+        elif arguments.get("icmp"):
+            protocol = "icmp"
+        else:
+            protocol = None
+
+        src_ports = arguments.get("<SRCPORTS>")
```

Comment:
I think the way it is now is cleaner - as we nest the arguments.get(...) calls it makes lines longer and harder to read.  I'd prefer to leave it this way.